sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -830,11 +839,105 @@ public void populateDisplayData(DisplayData.Builder builder) {
+      } catch (Exception e) {
+        throw new RuntimeException("Error while determining columns from table: " + inner.getTable());
+      }
+      List<String> databaseColumnList = getTableColumnList(connection);
+      fields = fields.stream().filter((field) -> {
+        return databaseColumnList.contains(field.getName().toLowerCase());
+      }).collect(Collectors.toList());
+      return fields;
+    }
+
+    private static void mapAccordingToSchema(Row element, PreparedStatement preparedStatement, Schema schema, List<Schema.Field> fields) {
+//      List<Schema.Field> fields = schema.getFields();
+      IntStream.range(0, fields.size()).forEach((index) -> {
+        JdbcUtil.setFieldPreparedStatement(element, preparedStatement, fields.get(index).getType(), index);
+      });
+    }
+
+    private List<String> getTableColumnList(Connection connection) {
+      try {
+        List<String> tableColumns = new ArrayList<>();
+        // get columns of database table
+        ResultSet resultSet = connection.getMetaData().getColumns(null, null, inner.getTable(), null);
+        while (resultSet.next()) {
+          tableColumns.add(resultSet.getString("column_name").toLowerCase());
+        }
+        return tableColumns;
+      } catch (Exception e) {
+        throw new RuntimeException("Error while determining columns from table: " + inner.getTable());
+      }
+    }
+
+    private String generateStatement(String tableName, List<Schema.Field> fields) {
+      final StringBuilder statement = new StringBuilder();
+      statement.append("INSERT INTO ")
+              .append(tableName)
+              .append(" (");
+
+      IntStream.range(0, fields.size()).forEach((index) -> {
+        statement.append(fields.get(index).getName());
+        if (index < fields.size() -1) {
+          statement.append(", ");
+        }
+      });
+
+      statement.append(") VALUES (");
+
+      IntStream.range(0, fields.size()).forEach((index) -> {
+        if (index == fields.size() -1)
+          statement.append("?");
+        else
+          statement.append("?, ");
+      });
+      statement.append(")");
+      return statement.toString();
+    }
+
+    private class AutoGeneratedPreparedStatementSetter implements PreparedStatementSetter<T> {
+
+      private List<Schema.Field> fields;
+      private SerializableFunction<T, Row> toRowFn;
+      private Schema schema;
+
+      AutoGeneratedPreparedStatementSetter(Schema schema, List<Schema.Field> fields, SerializableFunction<T, Row> toRowFn) {
+        this.fields = fields;
+        this.toRowFn = toRowFn;
+        this.schema = schema;
+      }
+
+      @Override
+      public void setParameters(T element, PreparedStatement preparedStatement) throws Exception {
+          Row row = (element instanceof Row) ? (Row) element : toRowFn.apply(element);
+          mapAccordingToSchema(row, preparedStatement, schema, fields);
+      }
```

Comment:
@charithe In the last commit we have added support of writing any type of data to JDBC. We use the toRowFunction available in the PCollection to convert T to Row.
See the test `testWriteWithoutPreparedStatementAndNonRowType()` which writes KV to JDBC.